src/main/java/com/yahoo/bullet/bql/query/TypeChecker.java

```diff
@@ -0,0 +1,386 @@
+/*
+ *  Copyright 2020, Yahoo Inc.
+ *  Licensed under the terms of the Apache License, Version 2.0.
+ *  See the LICENSE file associated with the project for terms.
+ */
+package com.yahoo.bullet.bql.query;
+
+import com.yahoo.bullet.bql.tree.BinaryExpressionNode;
+import com.yahoo.bullet.bql.tree.CastExpressionNode;
+import com.yahoo.bullet.bql.tree.ExpressionNode;
+import com.yahoo.bullet.bql.tree.FieldExpressionNode;
+import com.yahoo.bullet.bql.tree.ListExpressionNode;
+import com.yahoo.bullet.bql.tree.NAryExpressionNode;
+import com.yahoo.bullet.bql.tree.Node;
+import com.yahoo.bullet.bql.tree.UnaryExpressionNode;
+import com.yahoo.bullet.common.BulletError;
+import com.yahoo.bullet.query.expressions.Operation;
+import com.yahoo.bullet.querying.aggregations.grouping.GroupOperation;
+import com.yahoo.bullet.typesystem.Type;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+public class TypeChecker {
+    public static Optional<List<BulletError>> validateFieldType(FieldExpressionNode node, Type type, boolean hasIndexOrKey, boolean hasSubKey) {
+        if (Type.isUnknown(type)) {
+            return unknownError();
+        } else if (Type.isNull(type)) {
+            return makeError(node, "The field " + node.getField().getValue() + " does not exist in the schema.");
+        } else if ((hasIndexOrKey && !hasSubKey && !isCollection(type)) || (hasSubKey && !isComplex(type))) {
+            return makeError(node, "The subfield " + node + " is invalid since the field " + node.getField().getValue() + " has type: " + type);
+        }
+        return Optional.empty();
+    }
+
+    public static Type getFieldType(Type type, boolean hasIndexOrKey, boolean hasSubKey) {
+        if (hasSubKey) {
+            return type.getSubType().getSubType();
+        } else if (hasIndexOrKey) {
+            return type.getSubType();
+        }
+        return type;
+    }
+
+    public static Optional<List<BulletError>> validateListTypes(ListExpressionNode node, Set<Type> types) {
+        if (types.isEmpty()) {
+            return makeError(node, "Empty lists are currently not supported.");
+        }
+        if (types.contains(Type.UNKNOWN)) {
```

Comment:
Do we need to handle other UNKNOWN_* types?

Reply:
No. unknown in typechecker basically means "there's no type information for this expression so don't bother verifying." unknown lists and maps still have some type information (they're lists and maps), like you could still do SIZEOF an unknown list, though hopefully the user isnt adding unknown lists and maps to the schema